src/core/drive/preloader.ts

```diff
@@ -0,0 +1,48 @@
+import { expandURL } from "../url"
+import { Navigator } from "./navigator"
+import { PageSnapshot } from "./page_snapshot"
+import { SnapshotCache } from "./snapshot_cache"
+
+export interface PreloaderDelegate {
+  readonly navigator: Navigator
+}
+
+export class Preloader {
+  readonly delegate: PreloaderDelegate
+  readonly element: Element
+
+  constructor(delegate: PreloaderDelegate, element: Element) {
+    this.delegate = delegate
+    this.element = element
+  }
+
+  get snapshotCache(): SnapshotCache {
+    return this.delegate.navigator.view.snapshotCache
+  }
+
+  start() {
+    this.preloadOnLoadLinksForView(this.element)
+  }
+
+  preloadOnLoadLinksForView(element: Element) {
+    const links = element.querySelectorAll('a[data-turbo-preload="load"]')
+
+    for (const link of links) {
+      this.preloadURL(link)
+    }
+  }
+
+  preloadURL(link: Element) {
```

Comment:
If this method accepted [HTMLAnchorElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLAnchorElement) as a type, we could read the [link.href](https://developer.mozilla.org/en-US/docs/Web/API/HTMLAnchorElement/href) directly, without the need to call `expandURL` or `|| ""`.

Reply:
Thanks for the feedback @seanpdoyle I'll have a go at refining the types.

Reply:
https://github.com/hotwired/turbo/pull/552/commits/8e353855278743a3585fffc13b2bf0d27fbd51eb